### ISSUE TITLE
Assert in TryCompleteOrAbortAsync with Unix domain sockets

### ISSUE THREAD

**Issue by stephentoub** [dotnet/corefx#6806]
The following assert:
```
Unexpected oldState: Complete

   at System.Environment.GetStackTrace(Exception e, Boolean needFileInfo)
   at System.Environment.get_StackTrace()
   at System.Diagnostics.Debug.Assert(Boolean condition, String message, String detailMessage)
   at System.Diagnostics.Debug.Assert(Boolean condition, String message)
   at System.Net.Sockets.SocketAsyncContext.AsyncOperation.TryCompleteOrAbortAsync(SocketAsyncContext context, Boolean abort)
   at System.Net.Sockets.SocketAsyncContext.AsyncOperation.TryCompleteAsync(SocketAsyncContext context)
   at System.Net.Sockets.SocketAsyncContext.OperationQueue`1.Complete(SocketAsyncContext context)
   at System.Net.Sockets.SocketAsyncContext.HandleEvents(SocketEvents events)
   at System.Net.Sockets.SocketAsyncEngine.EventLoop()
   at System.Threading.Tasks.Task.Execute()
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot)
   at System.Threading.Tasks.Task.ExecuteEntry(Boolean bPreventDoubleExecution)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)Aborted (core dumped)
```
gets triggers by this repro:
```C#
using System;
using System.IO;
using System.IO.Pipes;
using System.Net.Http;
using System.Diagnostics;
using System.Threading.Tasks;
using System.Text;
using System.Net;
using System.Net.Sockets;

public class Program
{
    public static void Main()
    {
        using (Socket server = new Socket(AddressFamily.Unix, SocketType.Stream, ProtocolType.Unspecified))
        using (Socket client = new Socket(AddressFamily.Unix, SocketType.Stream, ProtocolType.Unspecified))
        {
            const int Iters = 100; // Increase this value
            const int Chunk = 1024;
            byte[] sendData = new byte[Chunk * Iters];
            byte[] receiveData = new byte[sendData.Length];
            new Random().NextBytes(sendData);

            string path = "./example.sock";
            File.Delete(path);
            server.Bind(new UnixDomainSocketEndPoint(path));
            server.Listen(1);

            Task<Socket> acceptTask = server.AcceptAsync();
            client.Connect(new UnixDomainSocketEndPoint(path));
            acceptTask.Wait();
            Socket accepted = acceptTask.Result;

            Task[] writes = new Task[Iters];
            Task<int>[] reads = new Task<int>[Iters];
            for (int i = 0; i < Iters; i++)
                writes[i] = client.SendAsync(new ArraySegment<byte>(sendData, i * Chunk, Chunk), SocketFlags.None);
            for (int i = 0; i < Iters; i++)
                reads[i] = accepted.ReceiveAsync(new ArraySegment<byte>(receiveData, i * Chunk, Chunk), SocketFlags.None);
            Task.WaitAll(writes);
            Task.WaitAll(reads);

            for (int i = 0; i < sendData.Length; i++)
                if (receiveData[i] != sendData[i])
                    throw new Exception($"Different at {i}");

            Console.WriteLine("Done");
        }
    }
}

// Copied verbatim from https://github.com/dotnet/corefx/blob/master/src/Common/src/System/Net/Sockets/UnixDomainSocketEndPoint.cs
namespace System.Net.Sockets
{
    /// <summary>Represents a Unix Domain Socket endpoint as a path.</summary>
    public sealed class UnixDomainSocketEndPoint : EndPoint
    {
        private const int MaxPathLength = 92;   // sockaddr_un.sun_path at http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/sys_un.h.html
        private const int PathOffset = 2;       // = offsetof(struct sockaddr_un, sun_path). It's the same on Linux and OSX
        private const int MaxSocketAddressSize = PathOffset + MaxPathLength;
        private const int MinSocketAddressSize = PathOffset + 2; // +1 for one character and +1 for \0 ending
        private const AddressFamily EndPointAddressFamily = AddressFamily.Unix;

        private static readonly Encoding s_pathEncoding = Encoding.UTF8;
        private readonly string _path;
        private readonly byte[] _encodedPath;

        public UnixDomainSocketEndPoint(string path)
        {
            if (path == null)
            {
                throw new ArgumentNullException(nameof(path));
            }

            if (path.Length == 0 || s_pathEncoding.GetByteCount(path) >= MaxPathLength)
            {
                throw new ArgumentOutOfRangeException(nameof(path));
            }

            _path = path;
            _encodedPath = s_pathEncoding.GetBytes(_path);
        }

        internal UnixDomainSocketEndPoint(SocketAddress socketAddress)
        {
            if (socketAddress == null)
            {
                throw new ArgumentNullException(nameof(socketAddress));
            }

            if (socketAddress.Family != EndPointAddressFamily || 
                socketAddress.Size > MaxSocketAddressSize)
            {
                throw new ArgumentOutOfRangeException(nameof(socketAddress));
            }

            if (socketAddress.Size >= MinSocketAddressSize)
            {
                _encodedPath = new byte[socketAddress.Size - PathOffset];
                for (int index = 0; index < socketAddress.Size - PathOffset; index++)
                {
                    _encodedPath[index] = socketAddress[PathOffset + index];
                }

                _path = s_pathEncoding.GetString(_encodedPath, 0, _encodedPath.Length);
            }
            else
            {
                // Empty path may be used by System.Net.Socket logging.
                _encodedPath = Array.Empty<byte>();
                _path = string.Empty;
            }
        }

        public string Path { get { return _path; } }

        public override AddressFamily AddressFamily { get { return EndPointAddressFamily; } }

        public override SocketAddress Serialize()
        {
            var result = new SocketAddress(AddressFamily.Unix, MaxSocketAddressSize);

            // Ctor has already checked that PathOffset + _encodedPath.Length < MaxSocketAddressSize
            for (int index = 0; index < _encodedPath.Length; index++)
            {
                result[PathOffset + index] = _encodedPath[index];
            }
            result[PathOffset + _encodedPath.Length] = 0; // The path must be ending with \0

            return result;
        }

        public override EndPoint Create(SocketAddress socketAddress)
        {
            return new UnixDomainSocketEndPoint(socketAddress);
        }

        public override string ToString()
        {
            return Path;
        }
    }
}
```
As you increase Iters to increase the number of concurrent operations, the change of hitting the assert gets much larger.

**Comment by ericeil** [comment]
The assert failure is addressed by #6814.  This code still fails with that change, though, due to an ordering issue; a call to ReceiveAsync (for example) that races with the completion of a previous ReceiveAsync call, can "jump to the head of the line" and "steal" the previous call's data.  This is because we aggressively call into the OS to receive on the socket, even if we've queued prior receives.  Instead, we need to check the queue, and if other operations are waiting, we need to simply enqueuer this one, rather than trying to complete it.

**Comment by ericeil** [comment]
Closing this issue, as the assert is now fixed.  Opened #6832 to cover the ordering issue.
